src/Eloquent/DocumentModel.php

```diff
@@ -0,0 +1,749 @@
+
+use function array_key_exists;
+use function array_keys;
+use function array_merge;
+use function array_unique;
+use function array_values;
+use function class_basename;
+use function count;
+use function date_default_timezone_get;
+use function explode;
+use function func_get_args;
+use function in_array;
+use function is_array;
+use function is_numeric;
+use function is_string;
+use function ltrim;
+use function method_exists;
+use function sprintf;
+use function str_contains;
+use function str_starts_with;
+use function strcmp;
+use function var_export;
+
+trait DocumentModel
+{
+    use HybridRelations;
+    use EmbedsRelations;
+
+    /**
+     * The parent relation instance.
+     */
+    private Relation $parentRelation;
+
+    /**
+     * List of field names to unset from the document on save.
+     *
+     * @var array{string, true}
+     */
+    private array $unset = [];
+
+    /**
+     * Custom accessor for the model's id.
+     *
+     * @param  mixed $value
+     *
+     * @return mixed
+     */
+    public function getIdAttribute($value = null)
+    {
+        // If we don't have a value for 'id', we will use the MongoDB '_id' value.
+        // This allows us to work with models in a more sql-like way.
+        if (! $value && array_key_exists('_id', $this->attributes)) {
+            $value = $this->attributes['_id'];
+        }
+
+        // Convert ObjectID to string.
+        if ($value instanceof ObjectID) {
+            return (string) $value;
+        }
+
+        if ($value instanceof Binary) {
+            return (string) $value->getData();
+        }
+
+        return $value;
+    }
+
+    /** @inheritdoc */
+    public function getQualifiedKeyName()
+    {
+        return $this->getKeyName();
+    }
+
+    /** @inheritdoc */
+    public function fromDateTime($value)
+    {
+        // If the value is already a UTCDateTime instance, we don't need to parse it.
+        if ($value instanceof UTCDateTime) {
+            return $value;
+        }
+
+        // Let Eloquent convert the value to a DateTime instance.
+        if (! $value instanceof DateTimeInterface) {
+            $value = parent::asDateTime($value);
+        }
+
+        return new UTCDateTime($value);
+    }
+
+    /** @inheritdoc */
+    protected function asDateTime($value)
+    {
+        // Convert UTCDateTime instances to Carbon.
+        if ($value instanceof UTCDateTime) {
+            return Date::instance($value->toDateTime())
+                ->setTimezone(new DateTimeZone(date_default_timezone_get()));
+        }
+
+        return parent::asDateTime($value);
+    }
+
+    /** @inheritdoc */
+    public function getDateFormat()
+    {
+        return $this->dateFormat ?: 'Y-m-d H:i:s';
+    }
+
+    /** @inheritdoc */
+    public function freshTimestamp()
+    {
+        return new UTCDateTime(Date::now());
+    }
+
+    /** @inheritdoc */
+    public function getTable()
+    {
+        return $this->collection ?? parent::getTable();
```

Comment:
This property is not defined in the trait, otherwise it's allowed to redefine in the class that uses the trait.